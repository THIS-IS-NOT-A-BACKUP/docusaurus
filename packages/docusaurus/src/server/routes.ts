/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import logger from '@docusaurus/logger';
import {normalizeUrl} from '@docusaurus/utils';
import type {RouteConfig, ReportingSeverity} from '@docusaurus/types';

// Recursively get the final routes (routes with no subroutes)
export function getAllFinalRoutes(routeConfig: RouteConfig[]): RouteConfig[] {
  function getFinalRoutes(route: RouteConfig): RouteConfig[] {
    return route.routes ? route.routes.flatMap(getFinalRoutes) : [route];
  }
  return routeConfig.flatMap(getFinalRoutes);
}

export function handleDuplicateRoutes(
  routes: RouteConfig[],
  onDuplicateRoutes: ReportingSeverity,
): void {
  if (onDuplicateRoutes === 'ignore') {
    return;
  }
  const allRoutes: string[] = getAllFinalRoutes(routes).map(
    (routeConfig) => routeConfig.path,
  );
  const seenRoutes = new Set<string>();
  const duplicatePaths = allRoutes.filter((route) => {
    if (seenRoutes.has(route)) {
      return true;
    }
    seenRoutes.add(route);
    return false;
  });
  if (duplicatePaths.length > 0) {
    logger.report(
      onDuplicateRoutes,
    )`Duplicate routes found!${duplicatePaths.map(
      (duplicateRoute) =>
        logger.interpolate`Attempting to create page at url=${duplicateRoute}, but a page already exists at this route.`,
    )}
This could lead to non-deterministic routing behavior.`;
  }
}

/**
 * Old stuff
 * As far as I understand, this is what permits to SSG the 404.html file
 * This is rendered through the catch-all ComponentCreator("*") route
 * Note CDNs only understand the 404.html file by convention
 * The extension probably permits to avoid emitting "/404/index.html"
 */
const NotFoundRoutePath = '/404.html';

export function getRoutesPaths(
  routeConfigs: RouteConfig[],
  baseUrl: string,
): string[] {
  return [
    normalizeUrl([baseUrl, NotFoundRoutePath]),
    ...getAllFinalRoutes(routeConfigs).map((r) => r.path),
  ];
}
