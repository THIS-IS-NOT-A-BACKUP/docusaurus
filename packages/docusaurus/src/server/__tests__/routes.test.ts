/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import {loadRoutes, handleDuplicateRoutes} from '../routes';
import type {RouteConfig} from '@docusaurus/types';

describe('handleDuplicateRoutes', () => {
  const routes: RouteConfig[] = [
    {
      path: '/',
      component: '',
      routes: [
        {path: '/search', component: ''},
        {path: '/sameDoc', component: ''},
      ],
    },
    {
      path: '/',
      component: '',
      routes: [
        {path: '/search', component: ''},
        {path: '/sameDoc', component: ''},
        {path: '/uniqueDoc', component: ''},
      ],
    },
    {
      path: '/',
      component: '',
    },
    {
      path: '/',
      component: '',
    },
    {
      path: '/',
      component: '',
    },
  ];
  it('works', () => {
    expect(() => {
      handleDuplicateRoutes(routes, 'throw');
    }).toThrowErrorMatchingSnapshot();
    const consoleMock = jest.spyOn(console, 'log').mockImplementation(() => {});
    handleDuplicateRoutes(routes, 'ignore');
    expect(consoleMock).toBeCalledTimes(0);
  });
});

describe('loadRoutes', () => {
  it('loads nested route config', async () => {
    const nestedRouteConfig: RouteConfig = {
      component: '@theme/DocPage',
      path: '/docs:route',
      modules: {
        docsMetadata: 'docs-b5f.json',
      },
      routes: [
        {
          path: '/docs/hello',
          component: '@theme/DocItem',
          exact: true,
          modules: {
            content: 'docs/hello.md',
            metadata: 'docs-hello-da2.json',
          },
          sidebar: 'main',
        },
        {
          path: 'docs/foo/baz',
          component: '@theme/DocItem',
          modules: {
            content: 'docs/foo/baz.md',
            metadata: 'docs-foo-baz-dd9.json',
          },
          sidebar: 'secondary',
          'key:a': 'containing colon',
          "key'b": 'containing quote',
          'key"c': 'containing double quote',
          'key,d': 'containing comma',
          字段: 'containing unicode',
        },
      ],
    };
    await expect(
      loadRoutes([nestedRouteConfig], '/', 'ignore'),
    ).resolves.toMatchSnapshot();
  });

  it('loads flat route config', async () => {
    const flatRouteConfig: RouteConfig = {
      path: '/blog',
      component: '@theme/BlogListPage',
      exact: true,
      modules: {
        items: [
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
              query: {
                truncated: true,
              },
            },
            metadata: 'blog-2018-12-14-happy-first-birthday-slash-d2c.json',
          },
          {
            content: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            metadata: null,
          },
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            },
            metadata: null,
          },
        ],
      },
    };
    await expect(
      loadRoutes([flatRouteConfig], '/', 'ignore'),
    ).resolves.toMatchSnapshot();
  });

  it('rejects invalid route config', async () => {
    const routeConfigWithoutPath = {
      component: 'hello/world.js',
    } as RouteConfig;

    await expect(loadRoutes([routeConfigWithoutPath], '/', 'ignore')).rejects
      .toThrowErrorMatchingInlineSnapshot(`
            "Invalid route config: path must be a string and component is required.
            {\\"component\\":\\"hello/world.js\\"}"
          `);

    const routeConfigWithoutComponent = {
      path: '/hello/world',
    } as RouteConfig;

    await expect(loadRoutes([routeConfigWithoutComponent], '/', 'ignore'))
      .rejects.toThrowErrorMatchingInlineSnapshot(`
            "Invalid route config: path must be a string and component is required.
            {\\"path\\":\\"/hello/world\\"}"
          `);
  });

  it('loads route config with empty (but valid) path string', async () => {
    const routeConfig = {
      path: '',
      component: 'hello/world.js',
    } as RouteConfig;

    await expect(
      loadRoutes([routeConfig], '/', 'ignore'),
    ).resolves.toMatchSnapshot();
  });
});
