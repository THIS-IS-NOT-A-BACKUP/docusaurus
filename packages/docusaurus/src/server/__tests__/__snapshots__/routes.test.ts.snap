// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleDuplicateRoutes works 1`] = `
"Duplicate routes found!
- Attempting to create page at /search, but a page already exists at this route.
- Attempting to create page at /sameDoc, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
This could lead to non-deterministic routing behavior."
`;

exports[`loadRoutes loads flat route config 1`] = `
{
  "registry": {
    "__comp---theme-blog-list-pagea-6-a-7ba": {
      "loader": "() => import(/* webpackChunkName: '__comp---theme-blog-list-pagea-6-a-7ba' */ '@theme/BlogListPage')",
      "modulePath": "@theme/BlogListPage",
    },
    "content---blog-0-b-4-09e": {
      "loader": "() => import(/* webpackChunkName: 'content---blog-0-b-4-09e' */ 'blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true')",
      "modulePath": "blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true",
    },
    "content---blog-7-b-8-fd9": {
      "loader": "() => import(/* webpackChunkName: 'content---blog-7-b-8-fd9' */ 'blog/2018-12-14-Happy-First-Birthday-Slash.md')",
      "modulePath": "blog/2018-12-14-Happy-First-Birthday-Slash.md",
    },
    "metadata---blog-0-b-6-74c": {
      "loader": "() => import(/* webpackChunkName: 'metadata---blog-0-b-6-74c' */ 'blog-2018-12-14-happy-first-birthday-slash-d2c.json')",
      "modulePath": "blog-2018-12-14-happy-first-birthday-slash-d2c.json",
    },
  },
  "routesChunkNames": {
    "/blog-599": {
      "__comp": "__comp---theme-blog-list-pagea-6-a-7ba",
      "items": [
        {
          "content": "content---blog-0-b-4-09e",
          "metadata": "metadata---blog-0-b-6-74c",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
      ],
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/blog',
    component: ComponentCreator('/blog', '599'),
    exact: true
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "/blog",
  ],
}
`;

exports[`loadRoutes loads nested route config 1`] = `
{
  "registry": {
    "__comp---theme-doc-item-178-a40": {
      "loader": "() => import(/* webpackChunkName: '__comp---theme-doc-item-178-a40' */ '@theme/DocItem')",
      "modulePath": "@theme/DocItem",
    },
    "__comp---theme-doc-page-1-be-9be": {
      "loader": "() => import(/* webpackChunkName: '__comp---theme-doc-page-1-be-9be' */ '@theme/DocPage')",
      "modulePath": "@theme/DocPage",
    },
    "content---docs-foo-baz-8-ce-61e": {
      "loader": "() => import(/* webpackChunkName: 'content---docs-foo-baz-8-ce-61e' */ 'docs/foo/baz.md')",
      "modulePath": "docs/foo/baz.md",
    },
    "content---docs-helloaff-811": {
      "loader": "() => import(/* webpackChunkName: 'content---docs-helloaff-811' */ 'docs/hello.md')",
      "modulePath": "docs/hello.md",
    },
    "docsMetadata---docs-routef-34-881": {
      "loader": "() => import(/* webpackChunkName: 'docsMetadata---docs-routef-34-881' */ 'docs-b5f.json')",
      "modulePath": "docs-b5f.json",
    },
    "metadata---docs-foo-baz-2-cf-fa7": {
      "loader": "() => import(/* webpackChunkName: 'metadata---docs-foo-baz-2-cf-fa7' */ 'docs-foo-baz-dd9.json')",
      "modulePath": "docs-foo-baz-dd9.json",
    },
    "metadata---docs-hello-956-741": {
      "loader": "() => import(/* webpackChunkName: 'metadata---docs-hello-956-741' */ 'docs-hello-da2.json')",
      "modulePath": "docs-hello-da2.json",
    },
  },
  "routesChunkNames": {
    "/docs/hello-44b": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "content": "content---docs-helloaff-811",
      "metadata": "metadata---docs-hello-956-741",
    },
    "/docs:route-52d": {
      "__comp": "__comp---theme-doc-page-1-be-9be",
      "docsMetadata": "docsMetadata---docs-routef-34-881",
    },
    "docs/foo/baz-070": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "content": "content---docs-foo-baz-8-ce-61e",
      "metadata": "metadata---docs-foo-baz-2-cf-fa7",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/docs:route',
    component: ComponentCreator('/docs:route', '52d'),
    routes: [
      {
        path: '/docs/hello',
        component: ComponentCreator('/docs/hello', '44b'),
        exact: true,
        sidebar: \\"main\\"
      },
      {
        path: 'docs/foo/baz',
        component: ComponentCreator('docs/foo/baz', '070'),
        sidebar: \\"secondary\\",
        \\"key:a\\": \\"containing colon\\",
        \\"key'b\\": \\"containing quote\\",
        \\"key\\\\\\"c\\": \\"containing double quote\\",
        \\"key,d\\": \\"containing comma\\",
        字段: \\"containing unicode\\"
      }
    ]
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "/docs/hello",
    "docs/foo/baz",
  ],
}
`;

exports[`loadRoutes loads route config with empty (but valid) path string 1`] = `
{
  "registry": {
    "__comp---hello-world-jse-0-f-b6c": {
      "loader": "() => import(/* webpackChunkName: '__comp---hello-world-jse-0-f-b6c' */ 'hello/world.js')",
      "modulePath": "hello/world.js",
    },
  },
  "routesChunkNames": {
    "-b2a": {
      "__comp": "__comp---hello-world-jse-0-f-b6c",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '',
    component: ComponentCreator('', 'b2a')
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "",
  ],
}
`;
